### PR TITLE
Convert errors to exceptions; move to supporting PHP 7.0+ only (for v2 branch)

### LINKED ISSUE #93: Replace PHP errors with PHP exceptions
PHP supports exceptions and it is much easier to deal with an exception than an error.

### PR DESCRIPTION
**Pull request by paragonie-scott**
Seeing if Travis CI passes it first. I probably need to update the tests to reflect this API change.

Closes #93, #107.

### CODE REVIEW

**Comment by paragonie-scott** [comment]
Crap, I just remembered that this sends an email with every commit. I'm going to get a proper IDE set up and then I'll make one last commit when this is ready to rock and roll.

**Comment by Jan-E** [comment]
Why bother? People that are watching github-repo's are used to getting a lot of emails. Separate commits makes it easier to track history. Can't you get your current IDE to push to github only if you explicitly tell it to do such? Then all commits will invoke only one email.

**Comment by paragonie-scott** [comment]
> Can't you get your current IDE to push to github only if you explicitly tell it to do such? Then all commits will invoke only one email.

I was using Github to trigger Travis CI builds to make sure there were no version-specific breaks.

**Comment by paragonie-scott** [comment]
@jedisct1 I believe this PR is ready now.

**Comment by Jan-E** [comment]
> I was using Github to trigger Travis CI builds to make sure there were no version-specific breaks.

If you want that, no IDE can prevent the emails evry time you trigger Travis. Don't  bother. BTW. Congrats with the outcome of https://wiki.php.net/rfc/libsodium

File: libsodium.c [hunk 1]
```diff
@@ -1522,7 +1524,7 @@ PHP_FUNCTION(sodium_crypto_sign_verify_detached)
         return;
     }
     if (signature_len != crypto_sign_BYTES) {
-        zend_throw_exception(zend_ce_exception, 
+        zend_throw_exception(zend_ce_exception,
```
Comment (paragonie-scott) [inline libsodium.c]: I'm not sure why Atom decided to fix that all of the sudden.

**Comment by paragonie-scott** [comment]
I think there might be some other specific code that only exists for PHP 5 compat? But I'm afraid to rip anything out and make a huge mess of the code.

**Comment by nikic** [comment]
Wondering if secret leakage through arguments captured in exception backtraces are a concern for this library. Of course this is only a problem in malconfigured environments that print exceptions in production, but for this library in particular it still might be a concern.

**Comment by jedisct1** [comment]
This is definitely a concern. Misconfigured environments printing exceptions in production are unfortunately very common.
According to what people build using these functions, anything, including nonces and public keys, can potentially leak sensitive information.

**Comment by nikic** [comment]
Yeah, agree. I don't think we have a native ability to disable arg collection for exception traces, so the simplest way to avoid this is probably to just create the exception normally and drop all the argument information after the fact.

**Comment by nikic** [comment]
Another question: Should this be throwing `Exception` or rather something more specific (like `SodiumException`)?

**Comment by paragonie-scott** [comment]
Exceptions are the best option, however: https://paragonie.com/blog/2017/02/cryptographically-secure-php-development#exceptions

> Another question: Should this be throwing `Exception` or rather something more specific (like `SodiumException`)?

I'd prefer a more specific exception but I'm not sure how to make that change in C.

**Comment by paragonie-scott** [comment]
Superseded by #113.

**Pull request closed without merging**